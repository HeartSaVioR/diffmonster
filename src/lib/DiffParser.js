import { Diff2Html } from 'diff2html';

export const LineType = {
  CONTEXT: 'd2h-cntx',
  ADDITION: 'd2h-ins',
  DELETION: 'd2h-del',
  NOEOL: '\\', // bye
};

export function parseDiff(diff) {
  // generate PR files API-like object https://developer.github.com/v3/pulls/#list-pull-requests-files
  const parsed = Diff2Html.getJsonFromDiff(diff);
  parsed.forEach(file => {
    file.filename = file.isDeleted ? file.oldName : file.newName;
    file.sha = file.isDeleted ? file.checksumBefore : file.checksumAfter;
    if (file.isDeleted) {
      file.status = 'removed';
    } else if (file.isNew || file.isCopy) {
      file.status = 'added';
    } else if (file.isRename) {
      file.status = 'renamed';
      file.previous_filename = file.oldName;
    }
    let position = 0;
    file.blocks.forEach(block => {
      position++;
      block.lines.forEach(line => {
        line.position = position++;
        line.content = line.content.slice(1);
      });
    });
  });
  return parsed;
}
