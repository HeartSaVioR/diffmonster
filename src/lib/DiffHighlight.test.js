import { parseDiff } from './DiffParser';
import { highlightDiff } from './DiffHighlight';

function parse(block) {
  return parseDiff(`diff --git a b
--- a
+++ b
${block}`)[0].blocks[0];
}

describe('DiffHighlight', () => {
  it('highlights the beginning of a line', () => {
    const patch = parse(`@@ -1,3 +1,3 @@
 aaa
-bbb
+0bb
 ccc`);
    expect(highlightDiff(patch)).toMatchSnapshot();
  });

  it('highlights the end of a line', () => {
    const patch = parse(`@@ -1,3 +1,3 @@
 aaa
-bbb
+bb0
 ccc`);
    expect(highlightDiff(patch)).toMatchSnapshot();
  });

  it('highlights the middle of a line', () => {
    const patch = parse(`@@ -1,3 +1,3 @@
 aaa
-bbb
+b0b
 ccc`);
    expect(highlightDiff(patch)).toMatchSnapshot();
  });

  it('does not highlight whole line', () => {
    const patch = parse(`@@ -1,3 +1,3 @@
 aaa
-bbb
+000
 ccc`);
    expect(highlightDiff(patch)).toMatchSnapshot();
  });

  it('does not highlight boring pairs', () => {
    const patch = parse(`@@ -1,3 +1,3 @@
 aaa
-  
+     
 ccc`);
    const result = highlightDiff(patch);
    expect(result.some(({ content }) => content.some(part => part.highlight))).toBe(false);
  });

  it('does not highlight mismatched hunk size', () => {
    const patch = parse(`@@ -1,3 +1,3 @@
 aaa
-bbb
+b0b
+ccc`);
    const result = highlightDiff(patch);
    expect(result.some(({ content }) => content.some(part => part.highlight))).toBe(false);
  });
});
