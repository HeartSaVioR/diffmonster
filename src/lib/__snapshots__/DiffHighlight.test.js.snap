// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiffHighlight does not highlight whole line 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "content": "aaa",
        "highlight": false,
      },
    ],
    "newNumber": 1,
    "oldNumber": 1,
    "position": 1,
    "type": "d2h-cntx",
  },
  Object {
    "content": Array [
      Object {
        "content": "bbb",
        "highlight": false,
      },
    ],
    "newNumber": null,
    "oldNumber": 2,
    "position": 2,
    "type": "d2h-del",
  },
  Object {
    "content": Array [
      Object {
        "content": "000",
        "highlight": false,
      },
    ],
    "newNumber": 2,
    "oldNumber": null,
    "position": 3,
    "type": "d2h-ins",
  },
  Object {
    "content": Array [
      Object {
        "content": "ccc",
        "highlight": false,
      },
    ],
    "newNumber": 3,
    "oldNumber": 3,
    "position": 4,
    "type": "d2h-cntx",
  },
]
`;

exports[`DiffHighlight highlights the beginning of a line 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "content": "aaa",
        "highlight": false,
      },
    ],
    "newNumber": 1,
    "oldNumber": 1,
    "position": 1,
    "type": "d2h-cntx",
  },
  Object {
    "content": Array [
      Object {
        "content": "b",
        "highlight": true,
      },
      Object {
        "content": "bb",
        "highlight": false,
      },
    ],
    "newNumber": null,
    "oldNumber": 2,
    "position": 2,
    "type": "d2h-del",
  },
  Object {
    "content": Array [
      Object {
        "content": "0",
        "highlight": true,
      },
      Object {
        "content": "bb",
        "highlight": false,
      },
    ],
    "newNumber": 2,
    "oldNumber": null,
    "position": 3,
    "type": "d2h-ins",
  },
  Object {
    "content": Array [
      Object {
        "content": "ccc",
        "highlight": false,
      },
    ],
    "newNumber": 3,
    "oldNumber": 3,
    "position": 4,
    "type": "d2h-cntx",
  },
]
`;

exports[`DiffHighlight highlights the end of a line 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "content": "aaa",
        "highlight": false,
      },
    ],
    "newNumber": 1,
    "oldNumber": 1,
    "position": 1,
    "type": "d2h-cntx",
  },
  Object {
    "content": Array [
      Object {
        "content": "bb",
        "highlight": false,
      },
      Object {
        "content": "b",
        "highlight": true,
      },
    ],
    "newNumber": null,
    "oldNumber": 2,
    "position": 2,
    "type": "d2h-del",
  },
  Object {
    "content": Array [
      Object {
        "content": "bb",
        "highlight": false,
      },
      Object {
        "content": "0",
        "highlight": true,
      },
    ],
    "newNumber": 2,
    "oldNumber": null,
    "position": 3,
    "type": "d2h-ins",
  },
  Object {
    "content": Array [
      Object {
        "content": "ccc",
        "highlight": false,
      },
    ],
    "newNumber": 3,
    "oldNumber": 3,
    "position": 4,
    "type": "d2h-cntx",
  },
]
`;

exports[`DiffHighlight highlights the middle of a line 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "content": "aaa",
        "highlight": false,
      },
    ],
    "newNumber": 1,
    "oldNumber": 1,
    "position": 1,
    "type": "d2h-cntx",
  },
  Object {
    "content": Array [
      Object {
        "content": "b",
        "highlight": false,
      },
      Object {
        "content": "b",
        "highlight": true,
      },
      Object {
        "content": "b",
        "highlight": false,
      },
    ],
    "newNumber": null,
    "oldNumber": 2,
    "position": 2,
    "type": "d2h-del",
  },
  Object {
    "content": Array [
      Object {
        "content": "b",
        "highlight": false,
      },
      Object {
        "content": "0",
        "highlight": true,
      },
      Object {
        "content": "b",
        "highlight": false,
      },
    ],
    "newNumber": 2,
    "oldNumber": null,
    "position": 3,
    "type": "d2h-ins",
  },
  Object {
    "content": Array [
      Object {
        "content": "ccc",
        "highlight": false,
      },
    ],
    "newNumber": 3,
    "oldNumber": 3,
    "position": 4,
    "type": "d2h-cntx",
  },
]
`;
